mlir/include/mlir/Dialect/MIOpen/gridwise_convolution_implicit_gemm_util.h

```diff
@@ -68,3 +69,140 @@ static void populateSeqVal(const ArrayAttr &seqAttr,
     }
   }
 }
+
+static ConvolutionContext populateConvContext(miopen::GridwiseGemmOp &op) {
+  miopen::ConvOpType opType = ObtainConvDirection(op);
+
+  llvm::StringMap<std::pair<size_t, int64_t>> dimIndexVal;
+
+  auto filterLayoutAttr = op.getAttrOfType<ArrayAttr>("filter_layout");
+  auto filterDimensionAttr = op.getAttrOfType<ArrayAttr>("filter_dimension");
+  populateDimVal(filterLayoutAttr, filterDimensionAttr, dimIndexVal);
+  auto inputLayoutAttr = op.getAttrOfType<ArrayAttr>("input_layout");
+  auto inputDimensionAttr = op.getAttrOfType<ArrayAttr>("input_dimension");
+  populateDimVal(inputLayoutAttr, inputDimensionAttr, dimIndexVal);
+  auto outputLayoutAttr = op.getAttrOfType<ArrayAttr>("output_layout");
+  auto outputDimensionAttr = op.getAttrOfType<ArrayAttr>("output_dimension");
+  populateDimVal(outputLayoutAttr, outputDimensionAttr, dimIndexVal);
+
+  auto strideAttr = op.getAttrOfType<ArrayAttr>("strides");
+  llvm::SmallVector<int64_t, 0> strideVal;
+  populateSeqVal(strideAttr, strideVal);
+
+  auto dilationAttr = op.getAttrOfType<ArrayAttr>("dilations");
+  llvm::SmallVector<int64_t, 0> dilationVal;
+  populateSeqVal(dilationAttr, dilationVal);
+
+  auto paddingAttr = op.getAttrOfType<ArrayAttr>("padding");
+  llvm::SmallVector<int64_t, 0> paddingVal;
+  populateSeqVal(paddingAttr, paddingVal);
+
+  return {opType, dimIndexVal, strideVal, dilationVal, paddingVal};
+}
+
+struct InitParamsNonXDL : InitParams {
+  InitParamsNonXDL(int64_t mPerBlock, int64_t nPerBlock, int64_t kPerBlock,
+                   int64_t mPerThread, int64_t nPerThread, int64_t bSize)
+      : InitParams{mPerBlock, nPerBlock, kPerBlock}, gemmMPerThread(mPerThread),
+        gemmNPerThread(nPerThread), blockSize(bSize) {}
+  int64_t gemmMPerThread;
+  int64_t gemmNPerThread;
+  int64_t blockSize;
+};
+
+class PopulateParams : public PopulateParamsBase {
+private:
+  llvm::SmallVector<InitParamsNonXDL, 4> initParameters = {
+      // M/block N/block K/block M/thread N/thread blockSize
+      {128, 128, 8, 4, 4, 256},
+      {128, 64, 8, 4, 4, 128},
+      {64, 128, 4, 4, 4, 128},
+      {32, 32, 4, 2, 2, 64},
+  };
+
+  LogicalResult
+  calculateGemmABlockCopyPerformanceParameters(InitParamsNonXDL *param,
+                                               ConvolutionContext &ctx,
+                                               DerivedParams &derived) {
+    return calculateInputDerivedParams(param, param->blockSize, ctx, true,
+                                       derived);
+  }
+
+  LogicalResult
+  calculateGemmBBlockCopyPerformanceParameters(InitParamsNonXDL *param,
+                                               ConvolutionContext &ctx,
+                                               DerivedParams &derived) {
+
+    return calculateInputDerivedParams(param, param->blockSize, ctx, false,
+                                       derived);
+  }
+
+  int64_t calculateGemmCDestDataPerWrite(ConvolutionContext &ctx) {
+    int64_t outputVecLen = 0;
```

Comment:
move the implementation to `.cpp`, and document the semantic in the header?

Reply:
Will do once we reached at agreement at other review comments.

Reply:
There's already too much changes in the current PR. Will leave this as topic for the next PR.